crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -0,0 +1,180 @@
+use ruff_index::IndexVec;
+
+use crate::semantic_index::{
+    ast_ids::HasScopedExpressionId,
+    constraint::{Constraint, ConstraintNode, PatternConstraintKind},
+    AllConstraints,
+};
+use crate::semantic_index::{ScopedConstraintId, ScopedVisibilityConstraintId};
+use crate::types::{infer_expression_types, Truthiness};
+use crate::Db;
+
+const MAX_RECURSION_DEPTH: usize = 10;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) enum VisibilityConstraint {
+    AlwaysTrue,
+    VisibleIf(ScopedConstraintId),
+    VisibleIfNot(ScopedVisibilityConstraintId),
+    KleeneAnd(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+    KleeneOr(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct VisibilityConstraints {
+    constraints: IndexVec<ScopedVisibilityConstraintId, VisibilityConstraint>,
+}
+
+impl VisibilityConstraints {
+    pub(crate) fn new() -> Self {
+        Self {
+            constraints: IndexVec::from_iter([VisibilityConstraint::AlwaysTrue]),
+        }
+    }
+
+    pub(crate) fn add(&mut self, constraint: VisibilityConstraint) -> ScopedVisibilityConstraintId {
+        self.constraints.push(constraint)
+    }
+
+    pub(crate) fn add_merged(
+        &mut self,
+        a: ScopedVisibilityConstraintId,
+        b: ScopedVisibilityConstraintId,
+    ) -> ScopedVisibilityConstraintId {
+        match (&self.constraints[a], &self.constraints[b]) {
+            (_, VisibilityConstraint::VisibleIfNot(id)) if a == *id => {
+                ScopedVisibilityConstraintId::ALWAYS_TRUE
+            }
+            (VisibilityConstraint::VisibleIfNot(id), _) if *id == b => {
+                ScopedVisibilityConstraintId::ALWAYS_TRUE
+            }
+            _ => self.add(VisibilityConstraint::KleeneOr(a, b)),
+        }
+    }
+
+    pub(crate) fn add_sequence(
+        &mut self,
+        a: ScopedVisibilityConstraintId,
+        b: ScopedVisibilityConstraintId,
+    ) -> ScopedVisibilityConstraintId {
+        if a == ScopedVisibilityConstraintId::ALWAYS_TRUE {
+            b
```

Comment:
I thought about it. The way we construct the `And` constraints right now, this can never happen. But we should probably add the other path as well, in order not to confuse readers. It's not like that would be much slower, if at all (with branch prediction).